apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo/interceptor/OauthInterceptor.kt

```diff
@@ -0,0 +1,63 @@
+package com.apollographql.apollo.interceptor
+
+import com.apollographql.apollo.ApolloError
+import com.apollographql.apollo.ApolloException
+import com.apollographql.apollo.api.ApolloExperimental
+import com.apollographql.apollo.api.Response
+import com.apollographql.apollo.network.HttpExecutionContext
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.single
+
+@ApolloExperimental
+class OauthInterceptor(private val accessTokenProvider: AccessTokenProvider) : ApolloRequestInterceptor {
+  private fun <T> ApolloRequest<T>.withHeader(name: String, value: String): ApolloRequest<T> {
+    val httpRequestContext = executionContext[HttpExecutionContext.Request]
+        ?: HttpExecutionContext.Request(emptyMap())
+            .let {
+              it.copy(headers = it.headers + (name to value))
+            }
+
+    return ApolloRequest(
+        operation = operation,
+        scalarTypeAdapters = scalarTypeAdapters,
+        executionContext = executionContext + httpRequestContext
+    )
+  }
+
+  private suspend fun <T> proceedWithToken(request: ApolloRequest<T>, interceptorChain: ApolloInterceptorChain, token: String): Response<T> {
+    val newRequest = request.withHeader("Authorization", "Bearer $token")
+    return try {
+      interceptorChain.proceed(newRequest).single()
```

Comment:
Should it be a flow? As we are breaking a chain here. Upstream can potentially emit multiple responses via flow but we restrict only to accept single

Reply:
Is it `Upstream` or `Downstream` 🙃? I tend to think as the rest of the chain (the one we proceed with) as `Downstream` but maybe it's the opposite, I'm always getting confused by these kind of things.

In all cases, my reasoning here is be that the OauthInterceptor should be just before the network interceptor and that the network interceptor should emit only once? Is there a use case to put the OauthInterceptor before the normalized cache or anything else that can emit multiple values ?

Reply:
Well, we don't know for sure if network interceptor will emit only once, as it depends on implementation of NetworkTransport. Couple examples: GraphQL subsscriptions, http cache or it can be up to user provide custom implementation of NetworkTransport that will emit several responses.

Why not just return this:
```
return flow {
  request.withHeader("Authorization", "Bearer $token")
}. flatMapLatest(request ->
  interceptorChain.proceed(request)
}
```